pkg/proctree/process.go

```diff
@@ -0,0 +1,383 @@
+	return &processNode{
+		id:             id,
+		creationTime:   int(time.Now().UnixNano()),
+		threads:        make(map[int]*threadNode),
+		childProcesses: make(map[int]*processNode),
+		execInfo:       types.NewTimeSeries[procExecInfo](procExecInfo{}),
+	}, nil
+}
+
+// setGenInfoFromEventProtected fill the general info of the process (information
+// of the event given by every process received from it) in a way that it is filled
+// only once in an efficient way to reduce performance penalty.
+// This method uses the process lock to reduce unnecessary locking if not needed. Make sure to unlock
+// the lock before using this.
+func (p *processNode) setGenInfoFromEventProtected(event *trace.Event) {
+	p.genInfoLock.Do(
+		func() {
+			p.mutex.Lock()
+			p.fillGeneralInfo(
+				event.ProcessID,
+				event.UserID,
+				event.ContainerID,
+			)
+			p.mutex.Unlock()
+		},
+	)
+}
+
+// setGeneralInfoProtected is used to fill general information of a process
+// only once, but when the information cannot be retrieved from an event like
+// with setGenInfoFromEventProtected.
+// This method is not protected by locks.
+func (p *processNode) setGeneralInfoProtected(
+	nsPid int,
+	userId int,
+	containerId string,
+) {
+	p.genInfoLock.Do(
+		func() {
+			p.fillGeneralInfo(
+				nsPid,
+				userId,
+				containerId,
+			)
+		},
+	)
+}
+
+// fillGeneralInfo is a util function to fill general information in process node.
+// General information is an information that resides in every event from a process.
+// As such, this information should only be updated once (unless it is changeable).
+func (p *processNode) fillGeneralInfo(
+	nsPid int,
+	userId int,
+	containerId string,
+) {
+	p.setUserId(userId)
+	p.setContainerId(containerId)
+	p.setNsId(nsPid)
+}
+
+// addThreadBasic add the thread to the process node if it does not exist.
+// The function also tries to synchronize the thread exit time with the process if filled after
+// process exit.
+// This function *does not* add the thread to the process tree, so it should be added afterward.
+func (p *processNode) addThreadBasic(tid int) (*threadNode, error) {
+	t, exist := p.getThread(tid)
+	if exist {
+		return t, nil
+	}
+	var err error
+	t, err = newThreadNode(tid)
+	if err != nil {
+		return nil, err
+	}
+	// Update thread exit time to match process if process exited
+	t.setExitTime(p.getExitTime())
+	t.connectToProcess(p)
+	p.connectToThread(t)
+	return t, nil
+}
+
+// isAlive return if the process is alive at the given moment, according to existing information
+// of the node.
+func (p *processNode) isAlive(checkTime types.Timestamp) bool {
+	exitTime := p.getExitTime()
+	if exitTime == 0 {
+		return true
+	}
+	if checkTime >= exitTime {
+		return false
+	}
+	forkTime := p.getForkTime()
+	return checkTime >= forkTime
+}
+
+// setExitTime sets the process's exit time
+func (p *processNode) setExitTime(exitTime types.Timestamp) {
+	p.exitTime = exitTime
```

Comment:
So, your object is either thread-safe or it is not. Like in this case, for example. What happens if exitTime is being written while another thread is reading the integer? Lots of things might happen:

- if the integer (hardly the case) is split among 2 CPU cache lines, you might read just half of it (while the other half hasn't been updated, which would truncate the value)
- some threads might read one value, some others might read another (golang's memory model won't guarantee ordering)

If you want to keep getters that aren't protected by your big lock (mutex) than, at least, you've got to make them atomic values (with atomic loads/stores) so they can be used concurrently by simultaneous threads (OS or runtime's threads).

Reply:
The same happens with all other cases where you aren't protecting values by holding the mutex. The mutex is a "primitive" and will guarantee that all CPUs are coherent (because of barriers used by locking primitives) and all values are atomically changed (or transactions, when changing more than one field, are atomically started/ended). If you don't use a mutex, you *have* to use atomic Loads/Stores for thread-safety.

Reply:
I have added other comment somewhere saying that if the values are **immutable** then its ok to have getters without mutexes or atomic loads (but then there couldn't exist a setter for these values, they would have to be set in the New() function only and never changed.

Reply:
The current design puts the responsibility for locking and unlocking nodes on the Tree.
No method of the nodes lock the mutex, except for the `setGenInfoFromEventProtected` method.
To conclude - the process and thread nodes are not thread-safe. So, the comments here are not relevant in this context.
If you feel that some method of the process tree is not using mutexes as needed, than please comment there.